pact/Pact/Core/Serialise/LegacyPact.hs

```diff
@@ -1,325 +1,1084 @@
-        Just d -> return d
-        Nothing -> fail $ "Invalid decimal value: " ++ show s
-      decodeDecimal v = fail $ "Invalid decimal value: " ++ show v
-
-      decodeTime = JD.withObject "time" $ \o ->
-        (o JD..: "time" >>= mkTime pactISO8601Format) <|>
-        (o JD..: "timep" >>= mkTime highPrecFormat)
-
-      mkTime fmt v = case parseTime fmt v of
-              Just t -> return t
-              Nothing -> fail $ "Invalid time value, expected " ++ fmt
-
-      pactISO8601Format :: String
-      pactISO8601Format = "%Y-%m-%dT%H:%M:%SZ"
-
-      highPrecFormat :: String
-      highPrecFormat = "%Y-%m-%dT%H:%M:%S.%vZ"
-
-
-instance JD.FromJSON (Guard QualifiedName PactValue) where
-  parseJSON v = guardToPactValue <$> JD.parseJSON v
-
-
--- https://github.com/kadena-io/pact/blob/ba15517b56eba4fdaf6b2fbd3e5245eeedd0fc9f/src/Pact/Types/Term/Internal.hs#L802
-instance JD.FromJSON (Guard QualifiedName LegacyPactValue) where
-  parseJSON v = GKeyset <$> JD.parseJSON v
-    <|> GKeySetRef <$> parseRef v
-    <|> GUserGuard <$> JD.parseJSON v
-    <|> GCapabilityGuard <$> JD.parseJSON v
-    <|> GModuleGuard <$> JD.parseJSON v
-    where
-    parseRef = JD.withObject "KeySetRef" $ \o -> do
-      ref <- o JD..: "ksn"
-      ns <- o JD..:? "ns"
-      pure (KeySetName ref ns)
-
-
-instance JD.FromJSON (UserGuard QualifiedName LegacyPactValue) where
-  parseJSON = JD.withObject "UserGuard" $ \o ->
-    UserGuard
-      <$> o JD..: "fun"
-      <*> o JD..: "args"
-
-instance JD.FromJSON (CapabilityGuard QualifiedName LegacyPactValue) where
-  parseJSON = JD.withObject "CapabilityGuard" $ \o ->
-    CapabilityGuard
-      <$> o JD..: "cgName"
-      <*> o JD..: "cgArgs"
-      <*> o JD..:? "cgPactId"
-
-instance JD.FromJSON ModuleGuard where
-  parseJSON = JD.withObject "ModuleGuard" $ \o ->
-    ModuleGuard
-      <$> o JD..: "moduleName"
-      <*> o JD..: "name"
-
-fromLegacyPactValue :: LegacyPactValue -> PactValue
-fromLegacyPactValue = \case
-  Legacy_PLiteral ll -> case ll of
-    Legacy_LString t -> PLiteral (LString t)
-    Legacy_LInteger i -> PLiteral (LInteger i)
-    Legacy_LDecimal d -> PLiteral (LDecimal d)
-    Legacy_LBool b -> PLiteral (LBool b)
-    Legacy_LTime utc -> PTime utc
-  Legacy_PList v -> PList (fromLegacyPactValue <$> v)
-  Legacy_PObject o -> PObject (fromLegacyPactValue <$> o)
-  Legacy_PGuard g -> PGuard (guardToPactValue g)
-  Legacy_PModRef mref -> PModRef mref
-
-
-guardToPactValue :: Guard QualifiedName LegacyPactValue -> Guard QualifiedName PactValue
-guardToPactValue = \case
-  (GKeyset ks) -> GKeyset ks
-  (GKeySetRef kref) -> GKeySetRef kref
-  (GUserGuard (UserGuard n tm)) -> GUserGuard (UserGuard n (fromLegacyPactValue <$> tm))
-  (GCapabilityGuard (CapabilityGuard n args i)) -> GCapabilityGuard (CapabilityGuard n (fromLegacyPactValue <$> args) i)
-  (GModuleGuard mg) -> GModuleGuard mg
-  (GDefPactGuard dpg) -> GDefPactGuard dpg
```

Comment:
Do we have a canonical way that we're throwing invariants here, and where do they get picked up and communicated to the user by the interpreter? Some of them are inconsistent formats.